solr/contrib/ltr/src/java/org/apache/solr/ltr/search/LTRQParserPlugin.java

```diff
@@ -146,93 +149,114 @@ public LTRQParser(String qstr, SolrParams localParams, SolrParams params,
     @Override
     public Query parse() throws SyntaxError {
       // ReRanking Model
-      final String modelName = localParams.get(LTRQParserPlugin.MODEL);
-      if ((modelName == null) || modelName.isEmpty()) {
+      final String[] modelNames = localParams.getParams(LTRQParserPlugin.MODEL);
+      if ((modelNames == null) || modelNames.length==0 || modelNames[0].isEmpty()) {
         throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
             "Must provide model in the request");
       }
-
-      final LTRScoringModel ltrScoringModel = mr.getModel(modelName);
-      if (ltrScoringModel == null) {
-        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
-            "cannot find " + LTRQParserPlugin.MODEL + " " + modelName);
-      }
-
-      final String modelFeatureStoreName = ltrScoringModel.getFeatureStoreName();
-      final boolean extractFeatures = SolrQueryRequestContextUtils.isExtractingFeatures(req);
-      final String fvStoreName = SolrQueryRequestContextUtils.getFvStoreName(req);
-      // Check if features are requested and if the model feature store and feature-transform feature store are the same
-      final boolean featuresRequestedFromSameStore = (modelFeatureStoreName.equals(fvStoreName) || fvStoreName == null) ? extractFeatures:false;
-      if (threadManager != null) {
-        threadManager.setExecutor(req.getCore().getCoreContainer().getUpdateShardHandler().getUpdateExecutor());
-      }
-      final LTRScoringQuery scoringQuery = new LTRScoringQuery(ltrScoringModel,
-          extractEFIParams(localParams),
-          featuresRequestedFromSameStore, threadManager);
-
-      // Enable the feature vector caching if we are extracting features, and the features
-      // we requested are the same ones we are reranking with
-      if (featuresRequestedFromSameStore) {
-        scoringQuery.setFeatureLogger( SolrQueryRequestContextUtils.getFeatureLogger(req) );
+     
+      LTRScoringQuery[] rerankingQueries = new LTRScoringQuery[modelNames.length];
+      for (int i = 0; i < modelNames.length; i++) {
+        final LTRScoringQuery rerankingQuery;
+        if (!ORIGINAL_RANKING.equals(modelNames[i])) {
```

Comment:
The model name `originalRanking` is being given a special meaning here. I wonder if perhaps the differences between models could be transferred to the parameter names somehow (e.g. a new `original_model` parameter name alongside the existing `model` parameter name)? Then users could choose any model name they like, including for "original ranking" purposes.

https://github.com/apache/lucene-solr/pull/1705 proposes to factor out a `LTRQParserPlugin.newLTRScoringQuery` method (but I haven't yet explored fully w.r.t. how that might connect up here w.r.t. additional parameter names).

What do you think?

Reply:
Hi @cpoerschke , I think you are right, it is probably better to have a separate param to activate the interleaving with the original ranking.
I have just made the change and updated the tests.
They are green locally, I'll just wait for them to be green remotely as well

Reply:
Thinking a little bit more about it, even if we solve the "originalRanking" request through a special param, when returning the results, the [interleaving] transformer will need to present for the search results coming from the original ranking a value :

{
        "id":"GB18030TEST",
        "score":1.0005897,
        "[interleaving]":"OriginalRanking"},
      {
        "id":"UTF8TEST",
        "score":0.79656565,
        "[interleaving]":"myModel"}]
  }

This must follow the same format ( I wouldn't add a special field name different from [interleaing], this would complicate evaluators behaviours.

This means that in the end I guess we need to keep a special name to indicate the original ranking, and this name can't be used by the admin for any other model.
Given that, do you think is still beneficial to have a separate parameter?
Or just using the original approach, with the documented exclusivity of a special name, would look cleaner?

Reply:
Ah, good point about the special "OriginalRanking" also appearing in the "[interleaving]" transformer!

When using interleaving there's always at least two models to be interleaved, right?

The models could all be actual models or one of them could be the "OriginalRanking" pseudo-model.

I wonder if class inheritance might help us e.g.

```
class InterleavingLTRQParserPlugin extends LTRQParserPlugin
```

and (say)

```
<queryParser name="ltr" class="org.apache.solr.ltr.search.LTRQParserPlugin"/>
<queryParser name="iltr" class="org.apache.solr.ltr.search.InterleavingLTRQParserPlugin"/>
```

where

```
rq={!iltr model=myModelXYZ}
```

can be a convenience equivalent to

```
rq={!iltr model=OriginalRanking model=myModelXYZ}
```

i.e. if only one model is supplied then it is implied that the second model is the original ranking.

And if the special "OriginalRanking" name doesn't suit someone (either because they already have a real model that happens to be called "OriginalRanking" or because they would prefer a different descriptor in the "[interleaving]" transformer output) then something like

```
rq={!iltr originalRankingModel=noModel model=noModel model=someModel}
```

would allow them to call the "OriginalRanking" something else e.g. "noModel" instead.


We could even reject any "OriginalRanking" models that are actual models via something like

```
final String originalRankingModelName = localParams.get("originalRankingModel", "OriginalRanking" /* default */);
if (null != mr.getModel(originalRankingModelName)) {
  throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
                          "Found an actual '" + originalRankingModelName + "' model, please ...
}
```

----

However, the "[interleaving]" transformer still needs to know about the special name, hmm.


At the moment we have

```
public static boolean isOriginalRanking(LTRScoringQuery rerankingQuery){
  return rerankingQuery.getScoringModel() == null;
}
```

in LTRQParserPlugin and in LTRInterleavingTransformerFactory

```
if (isOriginalRanking(rerankingQuery)) {
  doc.addField(name, ORIGINAL_RANKING);
} else {
  doc.addField(name, rerankingQuery.getScoringModel().getName());
}
```

and if we gave LTRScoringQuery a getScoringModelName method

```
public String getScoringModelName() {
  return ltrScoringModel.getName();
}
```

then in LTRInterleavingTransformerFactory we could have

```
doc.addField(name, rerankingQuery.getScoringModelName());
```

if the

```
new LTRScoringQuery(null);
```

in LTRQParserPlugin becomes


```
new LTRScoringQuery(null) {
  @Override
  public String getScoringModelName() {
    return originalRankingModelName;
  }
};
```

but that wouldn't displace any other `isOriginalRanking(rerankingQuery)` calls, though if we had something like

```
final class OriginalRankingLTRScoringQuery extends LTRScoringQuery {
  private final String originalRankingModelName;
  public OriginalRankingLTRScoringQuery(String originalRankingModelName) {
    super(null /* LTRScoringModel */);
    this.originalRankingModelName = originalRankingModelName;
  }
  @Override
  public String getScoringModelName() {
    return this.originalRankingModelName;
  }
}
```

then `(rerankingQuery instanceof OriginalRankingLTRScoringQuery)` could replace `isOriginalRanking(rerankingQuery)` conceptually.

----

Just thinking out aloud here, haven't yet tried to turn any of the above into code.

Could you see inheritance (for LTRQParserPlugin and/or LTRScoringQuery) work potentially?

Reply:
So @cpoerschke , I made a step back with the latest commit, let me explain,
First of all, the approach you detailed can definitely work,
I don't want to undervalue your effort and contribution, I am actually quite grateful for the time you dedicated in reviewing my proposal, so I really appreciate that.
But let's aim to a Keep It Simple Stupid approach here (KISS) and avoid over-complication in a first release, if possible:

1) Apache Solr makes use of "special names" in various places :
-  "_DEFAULT_" to indicate the default feature/model store
-  "_version" to indicate a meta-field used for versioning/optimistic concurrency
ect
I notice the prefix and suffix '_' is used for them.
So let's use it  : "_OriginalRanking_" could be a fair name to identify the Apache Solr ranking pre-re-ranking.
I named it after the "OriginalScore" feature, already in place, happy to use a different name, but wouldn't spend to much time on it.
My proposal would be to use "_OriginalRanking_" or potentially "_OriginalScoreRanking_" if more readable

2) the user interact with the !ltr query parser as usual, and if he/she wants to use interleaving, he/she just passes a second model (yes, we are supporting only 2 interleaved model right now):
model=A model=B
If he/she wants to interleave with the original ranking:
model=A model=_OriginalRanking_
I think this is quite intuitive and easy to be used, quite explicit and following current Apache Solr nomenclature for special named items

3) what happens if a user had/wants to use "_OriginalRanking_" for one of his/her models?
In this unlikely event, a Jira issue can be raised and we design a possible solution at that time.
A possible simple approach could be to make the original ranking name parametric, so the user can specify the name in the solrconfig.xml
But I would delay this reasoning only if we get requests for that functionality from users

What do you think?
If we find a good compromise here, we could move the review to the re-scoring part, that is more delicate and we can make sure no mistake happened there (I definitely don't want to slow down normal re-ranking and I definitely want the interleaved re-ranking to be as fast as possible)

You find the latest changes in the latest commits, I will update the related documentation as well today.

Reply:
Hi @alessandrobenedetti! Apologies for not returning to here sooner.

Good point about Apache Solr already using "special names" in various places, sure, let's go with "_OriginalRanking_" as the special name then. Hmm, the UI treats underscore special here, so `"_OriginalRanking_"` is what's in the code currently.

> ... we could move the review to the re-scoring part, that is more delicate ...

"delicate" is an excellent word choice, thank you, I spent some pleasant time today continuing to learn more about the re-scoring part. Took a sort of outside-to-inside approach i.e. starting at `LTRQParserPlugin` and then looking at what changed and how the bits fit together.

If it's okay with you I'll push a branch to my fork and add comments/questions sometimes with commit links here, not quite sure if that might work, let's try?

Reply:
Thanks @cpoerschke ! Tomorrow I'll address all your comments and think about them!
My intent is to merge within few days, so hopefully we can close this soon!
Thank you again for your time and cntribution, much appreciated!